appscale/tools/appscale_tools.py

```diff
@@ -959,60 +955,50 @@ def upload_app(cls, options):
         'or a directory. Please try uploading either a tar.gz file, a zip ' \
         'file, or a directory.'.format(options.file))
 
-    app_language = AppEngineHelper.get_app_runtime_from_app_config(
-      file_location)
+    version = Version.from_source(options.file)
+
     if options.project:
-      if app_language == 'java':
+      if version.runtime == 'java':
         raise BadConfigurationException("AppScale doesn't support --project for"
           "Java yet. Please specify the application id in appengine-web.xml.")
 
-      app_id = options.project
-    else:
-      try:
-        app_id = AppEngineHelper.get_app_id_from_app_config(file_location)
-      except AppEngineConfigException as config_error:
-        AppScaleLogger.log(config_error)
-        if 'yaml' in str(config_error):
-          raise config_error
+      version.project_id = options.project
+
+    if version.project_id is None:
+      if version.configuration_type == 'app.yaml':
+        message = 'Specify --project or define "application" in your app.yaml'
+      else:
+        message = 'Define "application" in your appengine-web.xml'
 
-        # Java App Engine users may have specified their war directory. In that
-        # case, just move up one level, back to the app's directory.
-        file_location = file_location + os.sep + ".."
-        app_id = AppEngineHelper.get_app_id_from_app_config(file_location)
+      raise AppEngineConfigException(message)
```

Comment:
What's difference between `AppEngineConfigException` and `BadConfigurationException`?

Reply:
I think originally `AppEngineConfigException` was just for the app.yaml and appengine-web.xml while `BadConfigurationException` could be for other things like the AppScalefile. It doesn't look like we really stuck to that, though. I think they should be consolidated.

Reply:
In some cases, it would fail on both. In others, it would fail just on AppScale. I would be favor of changing that exception name, or even just replacing everything with `BadConfiguration`, but I think it should go in a separate pull request.